src/scheduler/types/Scheduler.py

```diff
@@ -377,18 +382,120 @@ def bucket_stats(vals):
             options += [(goodness, budget + stem.cost, info) for goodness, budget, info in best_by_budget if budget + stem.cost <= cost_threshold]
         results = scheduler_util.make_monotonic(options)
 
-        def extract_schedule(info_dct):
-            schedule = [info_dct['unit']]
-            while info_dct['prev'] is not None:
-                info_dct = info_dct['prev']
-                schedule.insert(0, info_dct['unit'])
-            return schedule
-
         best_result = results[0]
         if self.verbose > 1:
             print 'Best:', best_result[:2]
         # best_schedule = best_result[2]['schedule']
-        best_schedule = extract_schedule(best_result[2])
+        best_schedule = scheduler_util.extract_schedule(best_result[2])
+        if self.verbose > 1:
+            print 'Schedule cost:', scheduler_util.get_cost_schedule(best_schedule, self.model.layer_latencies, self.model.final_layer)
+        avg_metric = self.set_schedule_values(best_schedule)
+        return avg_metric
+
+    def stems_scheduler(self, cost_threshold):
+        # Currently: brute forces all stems, runs DP for each stem individually
+        #
+        # run-time: O(# of stems * num_apps * budget * num_frozen * fps
+        #
+        # the last num_frozen * fps could probably be amortized but this might
+        # be moot since # of stems grows ~O(num_frozen! fps!)
+        cost_benefits = self.get_cost_benefits()
+        target_fps_options = range(1, self.stream_fps + 1)
+        chokepoints = sorted(set(key for app in self.apps for key in app["accuracies"]))
+
+        func_init, agg_func = self.get_init_agg_func()
+
+        solutions = []
+        best_result = None
+        for k in range(1, min((len(target_fps_options), len(chokepoints), len(self.apps)))):
```

Comment:
Hm, seems to fail when chokepoints (or target_fps_options or apps) is of length == 1. However, we sometimes expect only one {chokepoint, target_fps} (i.e, no sharing or max sharing). I discovered this when merging master into this branch. I'll submit a PR with my merge that still has this bug in it.